From 4c38ccd8f63ec047bdeb5fd8800c731952e2ccdf Mon Sep 17 00:00:00 2001
From: Thierry Quervel <thierry.quervel@intel.com>
Date: Fri, 28 Sep 2012 10:32:54 +0200
Subject: [PORT FROM MAIN] Restore BT A2DP device availability as soon as possible

BZ: 50998

This patch resets the FORCE_NO_BT_A2DP flag to FORCE_NONE
when wired headset or headphone is unplugged (if no WIDI
device available).
This allows to swap directly on BT A2DP without transient
playback on speaker.

Feature: AudioComms, WiDi, A2DP
Category: Integration
Purpose: PV-Enabler
Origin: Internal
Upstream-candidate: no
Change-Id: If90c5bcfdd140e477051dcb4c871cac7c822ad78
Orig-Change-Id: Ib7c016c1c27aed8dc1a5b44702bf012e3c1c77b7
Signed-off-by: Thierry Quervel <thierry.quervel@intel.com>
---
 audio/AudioPolicyManagerBase.cpp |   17 +++++++++++++++++
 1 files changed, 17 insertions(+), 0 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index 0d0c195..76f2240 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -135,6 +135,23 @@ status_t AudioPolicyManagerBase::setDeviceConnectionState(audio_devices_t device
             // remove device from available output devices
             mAvailableOutputDevices = (audio_devices_t)(mAvailableOutputDevices & ~device);
 
+            // Change policy in the case of wsHS unplugging with BT A2DP connected:
+            // at that moment, the FORCE_NO_BT_A2DP flag is set; that unfortunately leads to select
+            // temporarily the speaker as output until setForceUse(FORCE_NONE) is called
+            // by AudioService.
+            // Check that BT A2DP connection is available, and that there is no WIDI
+            // device, otherwise keep the flag to prioritize WIDI over BT A2DP
+            if ((device == AudioSystem::DEVICE_OUT_WIRED_HEADSET) ||
+                (device == AudioSystem::DEVICE_OUT_WIRED_HEADPHONE)){
+                if (!isInCall() &&
+                    (mForceUse[AudioSystem::FOR_MEDIA] == AudioSystem::FORCE_NO_BT_A2DP) &&
+                    (getA2dpOutput() != 0) && !mA2dpSuspended &&
+                    !(mAvailableOutputDevices & AudioSystem::DEVICE_OUT_WIDI)) {
+                    ALOGD("Disabling priority over BT A2DP for MEDIA streams");
+                    mForceUse[AudioSystem::FOR_MEDIA] = AudioSystem::FORCE_NONE;
+                }
+            }
+
             checkOutputsForDevice(device, state, outputs);
             if (mHasA2dp && audio_is_a2dp_device(device)) {
                 // handle A2DP device disconnection
-- 
1.7.4.1

